component/src/main/java/org/wso2/extension/siddhi/execution/env/ResourceBatchWindowProcessor.java

```diff
@@ -0,0 +1,378 @@
+                           StreamEventCloner streamEventCloner) {
+        synchronized (this) {
+            long currentTime = siddhiAppContext.getTimestampGenerator().currentTime();
+            ComplexEventChunk<StreamEvent> outputStreamEventChunk = new ComplexEventChunk<StreamEvent>(true);
+            while (streamEventChunk.hasNext()) {
+                StreamEvent streamEvent = streamEventChunk.next();
+                StreamEvent clonedStreamEvent = streamEventCloner.copyStreamEvent(streamEvent);
+                if (!streamEvent.getType().equals(ComplexEvent.Type.TIMER)) {
+                    Object groupEventMapKey = groupKeyExpressionExecutor.execute(clonedStreamEvent);
+                    if (groupEventMap.get(groupEventMapKey) != null) {
+                        groupEventMap.get(groupEventMapKey).streamEventList.add(clonedStreamEvent);
+                    } else {
+                        List<StreamEvent> list = new ArrayList<>();
+                        list.add(clonedStreamEvent);
+                        groupEventMap.put(groupEventMapKey, new ResourceStreamEventList(list,
+                                clonedStreamEvent.getTimestamp() + timeInMilliSeconds));
+                        if (nextEmitTime == -1) {
+                            nextEmitTime = currentTime + timeInMilliSeconds;
+                            if (scheduler != null) {
+                                scheduler.notifyAt(nextEmitTime);
+                            }
+                        }
+                    }
+                }
+                for (Map.Entry<Object, ResourceStreamEventList> entry : groupEventMap.entrySet()) {
+                    windowLength = ResourceIdentifierStreamProcessor.getResourceCount(resourceName);
+                    if (entry.getValue().streamEventList.size() >= windowLength || entry.getValue().expiryTimestamp <
+                            currentTime) {
+                        //update current event chunk with event batch
+                        for (StreamEvent event : entry.getValue().streamEventList) {
+                            currentEventChunk.add(event);
+                        }
+                        groupEventMap.remove(entry.getKey());
+                        //update outputStreamEventChunk with expired events chunk
+                        if (outputExpectsExpiredEvents) {
+                            if (eventsToBeExpired.getFirst() != null) {
+                                while (eventsToBeExpired.hasNext()) {
+                                    StreamEvent expiredEvent = eventsToBeExpired.next();
+                                    expiredEvent.setTimestamp(currentTime);
+                                }
+                                outputStreamEventChunk.add(eventsToBeExpired.getFirst());
+                            }
+                            eventsToBeExpired.clear();
+                        }
```

Comment:
Can move the 
```
if (outputExpectsExpiredEvents) {
...
}
```
out of the for loop as we are only doing to do this once

Reply:
Fixes added: 6293c56c572504981545b93c605c818f4a04511f